components/zeekopenbsm/src/openbsmconsumer.cpp

```diff
@@ -0,0 +1,363 @@
+      break;
+    case AUT_SUBJECT32:
+    case AUT_SUBJECT64:
+    case AUT_SUBJECT32_EX:
+    case AUT_SUBJECT64_EX:
+      extractSubject(event, tok);
+      break;
+    case AUT_RETURN32:
+    case AUT_RETURN64:
+      extractReturn(event, tok);
+      break;
+    case AUT_SOCKINET32:
+    case AUT_SOCKINET128:
+      extractSocketInet(event, tok);
+      break;
+    }
+  } // for loop
+
+  return Status::success();
+}
+
+Status OpenbsmConsumer::parseMessage() {
+
+  Status status;
+  Event event;
+
+  u_char *buffer = nullptr;
+
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
+  }
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
+
+  auto event_id = 0;
+  auto bytesread = 0;
+  while (bytesread < reclen) {
+    if (au_fetch_tok(&tok, buffer + bytesread, reclen - bytesread) == -1) {
+      break;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+      event_id = tok.tt.hdr32.e_type;
+      break;
+    case AUT_HEADER32_EX:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER64:
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+  if (event_filter_list.find(event_id) == event_filter_list.end()) {
+    // Return early for unused event IDs.
+    return Status::success();
+  }
+
+  status = populateSocketEvent(event, tokens);
+
+  if (!status.succeeded()) {
+    d->logger.logMessage(IZeekLogger::Severity::Error, status.message());
+
+  } else {
+    std::lock_guard<std::mutex> lock(d->event_list_mutex);
+    d->event_list.push_back(std::move(event));
+    d->event_list_cv.notify_all();
+  }
+  return status;
+}
+
+Status OpenbsmConsumer::fetchEventsFromPipe() {
+
+  audit_pipe = fopen(AUDIT_PIPE_PATH, "r");
+  if (audit_pipe == nullptr) {
+    throw Status::failure("The auditpipe couldn't be opened.");
+  }
+  event_filter_list.insert(AUE_CONNECT);
+  event_filter_list.insert(AUE_BIND);
+  fd_set fdset;
+  struct timeval timeout {};
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 200000;
+
+  while (!d->terminate_producer) {
+    FD_ZERO(&fdset);
+    FD_SET(fileno(audit_pipe), &fdset);
+
+    int rc = select(FD_SETSIZE, &fdset, nullptr, nullptr, &timeout);
+
+    if (rc == 0) {
+      continue;
+    }
+    if (rc < 0) {
+      if (errno != EINTR) {
+        return Status::failure("Auditpipe cannot be read");
+      }
+      continue;
+    }
+    parseMessage();
+  }
+  return Status::success();
+}
+
+Status IOpenbsmConsumer::create(Ref &obj, IZeekLogger &logger,
+                                IZeekConfiguration &configuration) {
+  try {
+    obj.reset(new OpenbsmConsumer(logger, configuration));
+    return Status::success();
+
+  } catch (const std::bad_alloc &) {
+    return Status::failure("Memory allocation failure");
```

Comment:
Personally I find handling out-of-memory errors superfluous since we allocate in a lot of other places in the code where we never handle failures (e.g., we use `string`). I understand this is here for consistency with the rest of the code.

Reply:
Yes, you are correct. I am doing this just to be consistent with the current code style.